### ISSUE TITLE
Counter cache fix loads model, but some fields are not yet available and it fails

### ISSUE THREAD

**Issue by LucasArruda** [magnusvk/counter_culture#141]
First off, I would like to personally thank all the maintainers (and especially @magnusvk) for this great gem! We use it a lot in our team.

---

We took quite some time to understand why migrations were failing for us, but then it worked when executed again.

We have a model `Invitation` like so:

```ruby
#invitation.rb
  extend Enumerize
    enumerize :invitee_gender, in: UserProfile::GENDERS
…
```

When I run the migration that add the counter itself, I also fix the counters:

```ruby
class AddPlusOnesCountToInvitations < ActiveRecord::Migration
  def change
    add_column :invitations, :plus_ones_count, :integer, default: 0, null: false
   …
        Invitation.counter_culture_fix_counts batch_size: 100, skip_unsupported: true
   …
```

But this `migration` will fail because there is no `invitee_gender` in `Invitations` yet, as shown above:

```ruby
== 20160720201436 AddPlusOnesCountToInvitations: migrating ====================
-- add_column(:invitations, :plus_ones_count, :integer, {:default=>0, :null=>false})
   -> 0.3996s
rake aborted!
StandardError: An error has occurred, this and all later migrations canceled:

can't write unknown attribute `invitee_gender`
/home/sam/.rvm/gems/ruby-2.2.4/gems/activerecord-4.2.7/lib/active_record/attribute.rb:138:in `with_value_from_database'
...
```

Is there something I could do to improve this situation or a fix for it?

I thought of creating a `BogusInvitation` model so I could fake `Invitation` and its table, so `counter_culture` could do it's job.

What do you think?

Many thanks!

**Comment by magnusvk** [comment]
Hm, I'm not really sure I understand the problem here. You are trying to run `counter_culture_fix_counts` at a time when the model code already refers to an `invitee_gender` column but that column doesn't exist in the database? In that case why don't you just add the column?

Are you attempting to run old migrations? I would suggest doing a schema load instead. Alternatively, you would have to remove the reference to model code from the migration and run it from a Rake task or console instead.

But maybe I'm misunderstanding this?

**Comment by LucasArruda** [comment]
The thing is, when I run all migrations after a clean DB state, the migration that runs `counter_culture_fix_counts` will run upon `Invitation` model, which contains references to `invitee_gender` inside, such as `enumerize :invitee_gender, in: UserProfile::GENDERS`.

But, because that field is not yet in the DB (it's created by another migration, added later), we run into a `deadlock` situation where the resource is not available at the time of the creation.

I could inverse those migrations, but ended up creating a `FakeInvitation` model inside the `migration` and referencing table `invitations`, and it worked out fine.

**I was wondering if that was the best solution.** 

If I could run a plain SQL based on `counter_culture_fix_counts` but not depending on invoking `Invitation` model itself or something similar.

Thanks!

**Comment by magnusvk** [comment]
Yeah this is a general problem with using model code in your migrations.

In my opinion, it is a bad idea to run all your migrations on a clean DB state. There is really no need. It is more reliable and thus better to run `rake db:schema:load` instead. 

However, if you do want to keep running old migrations on an empty database, I would suggest not using model code in there. Is that call to fix the counts even necessary on an empty database?

**Comment by LucasArruda** [comment]
You have a point there.

The thing is the migrations are being run on staging, but not on production, of course. But `migrations` need to work on both.

Anyways, thanks for replying and helping. For now, it's working this way but I'll see if I can make it better.

Cheers!

**Comment by LucasArruda** [comment]
As a side note:
The migration was also necessary to run the `fix_counts` on production because a `counter` field has been added on that migration, and it needed to count back to all `records` that were there before.

First version, when I was not using `counter_culture` I updated the counters myself (plain `SQL`), so we didn't have this `Model` on `migration` issue. But we ended up using `counter_culture` because it provides a nicer way to name and manage counters than using Rails counters + plain `SQL` to migrate.

In that sense, I could have kept the plain `SQL` to update the created counter vs using `*_fix_counts`.
